### ISSUE TITLE
IPV6 does not work during docker build command

### ISSUE THREAD

**Issue by Firstyear** [docker/docker#16913]
During a docker build communication over IPv6 via fe80::1%eth0 does not work to the host machine. The built container however functions correctly. We can see that correctly ipv6 link local addresses do exist in /proc/net/if_inet6.

List the steps to reproduce the issue:
1. From a fedora:latest with docker run -i -t /bin/bash:
[root@c76f2b9d2d77]/# nmap 172.17.42.1 -p 3128

Starting Nmap 6.47 ( http://nmap.org ) at 2015-10-11 03:11 UTC
Nmap scan report for 172.17.42.1
Host is up (0.000065s latency).
PORT     STATE SERVICE
3128/tcp open  squid-http
MAC Address: 02:42:FC:D5:D8:71 (Unknown)

Nmap done: 1 IP address (1 host up) scanned in 0.13 seconds
[root@c76f2b9d2d77]/# nmap -6 fe80::1%eth0 -p 3128

Starting Nmap 6.47 ( http://nmap.org ) at 2015-10-11 03:11 UTC
Nmap scan report for gateway (fe80::1)
Host is up (0.00017s latency).
PORT     STATE SERVICE
3128/tcp open  squid-http
MAC Address: 02:42:FC:D5:D8:71 (Unknown)

Nmap done: 1 IP address (1 host up) scanned in 0.15 seconds

2:
Using the following Dockerfile:

FROM fedora:latest
RUN dnf install -y nmap
RUN cat /proc/net/if_inet6
RUN nmap 172.17.42.1 -p 3128
RUN nmap -6 fe80::1%eth0 -p 3128

With docker build -t test .

Describe the results you received:

docker build -t testcontainer:latest .
Sending build context to Docker daemon 2.048 kB
Step 0 : FROM fedora:latest
 ---> ded7cd95e059
Step 1 : RUN dnf install -y nmap
 ---> Running in 18661977613d
Last metadata expiration check performed 0:00:19 ago on Sun Oct 11 03:15:20 2015.
Dependencies resolved.
================================================================================
 Package           Arch           Version                 Repository       Size
================================================================================
Installing:
 libpcap           x86_64         14:1.7.3-1.fc22         updates         149 k
 nmap              x86_64         2:6.47-2.fc22           fedora          4.0 M
 nmap-ncat         x86_64         2:6.47-2.fc22           fedora          135 k

Transaction Summary
================================================================================
Install  3 Packages

Total download size: 4.3 M
Installed size: 17 M
Downloading Packages:
--------------------------------------------------------------------------------
Total                                           233 kB/s | 4.3 MB     00:18     
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Installing  : libpcap-14:1.7.3-1.fc22.x86_64                              1/3 
warning: Unable to get systemd shutdown inhibition lock
  Installing  : nmap-ncat-2:6.47-2.fc22.x86_64                              2/3 
  Installing  : nmap-2:6.47-2.fc22.x86_64                                   3/3 
  Verifying   : nmap-2:6.47-2.fc22.x86_64                                   1/3 
  Verifying   : nmap-ncat-2:6.47-2.fc22.x86_64                              2/3 
  Verifying   : libpcap-14:1.7.3-1.fc22.x86_64                              3/3 

Installed:
  libpcap.x86_64 14:1.7.3-1.fc22            nmap.x86_64 2:6.47-2.fc22           
  nmap-ncat.x86_64 2:6.47-2.fc22           

Complete!
 ---> aa84cdbacaed
Removing intermediate container 18661977613d
Step 2 : RUN cat /proc/net/if_inet6
 ---> Running in b6c8b1e812cb
00000000000000000000000000000001 01 80 10 80       lo
fe8000000000000000000242ac11000c bd 40 20 c0     eth0
fe800000000000000042acfffe11000c bd 40 20 c0     eth0
 ---> c5483d1e5e0b
Removing intermediate container b6c8b1e812cb
Step 3 : RUN nmap 172.17.42.1 -p 3128
 ---> Running in 89dab5398379

Starting Nmap 6.47 ( http://nmap.org ) at 2015-10-11 03:16 UTC
Nmap scan report for 172.17.42.1
Host is up (0.000053s latency).
PORT     STATE SERVICE
3128/tcp open  squid-http
MAC Address: 02:42:FC:D5:D8:71 (Unknown)

Nmap done: 1 IP address (1 host up) scanned in 0.13 seconds
 ---> fc208785eba9
Removing intermediate container 89dab5398379
Step 4 : RUN nmap -6 fe80::1%eth0 -p 3128
 ---> Running in f2c4bf3b8a1c

Starting Nmap 6.47 ( http://nmap.org ) at 2015-10-11 03:16 UTC
Failed to convert source address to presentation format!?!  Error: Address family not supported by protocol
QUITTING!
The command '/bin/sh -c nmap -6 fe80::1%eth0 -p 3128' returned a non-zero code: 1

Describe the results you expected:

IPv6 to work during docker build

Provide additional info you think is important:

This is not an interpretation issue with the Dockerfile and the % or other chars, as I picked this up after using an ADD file command to add a local proxy via ipv6 to a container. The following command that relied on the proxy then failed due to ipv6 not working.

In my case I am checking port 3128, but this could easily be anything on the host machine.

docker version
Client:
 Version:      1.8.2-fc22
 API version:  1.20
 Package Version: docker-1.8.2-1.gitf1db8f2.fc22.x86_64
 Go version:   go1.4.2
 Git commit:   f1db8f2/1.8.2
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2-fc22
 API version:  1.20
 Package Version: 
 Go version:   go1.4.2
 Git commit:   f1db8f2/1.8.2
 Built:        
 OS/Arch:      linux/amd64

docker info   
Containers: 5
Images: 73
Storage Driver: btrfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.3.0-0.rc4.git0.1.mbpv2.fc24.x86_64
Operating System: Fedora 22 (Twenty Two)
CPUs: 8
Total Memory: 15.46 GiB
Name: franky.ipa.example.com
ID: WGE2:VZWO:5DPK:4AV7:BVUB:2JGD:TVHN:S7HP:6AGD:CJ56:QMRJ:JSC4
Debug mode (server): true
File Descriptors: 25
Goroutines: 56
System Time: 2015-10-11T13:15:56.901268426+10:00
EventsListeners: 0
Init SHA1: 603c4f89aedd044c26f9b65e8f6af985190a4da3
Init Path: /usr/libexec/docker/dockerinit
Docker Root Dir: /var/lib/docker

Linux franky.ipa.example.com 4.3.0-0.rc4.git0.1.mbpv2.fc24.x86_64 #1 SMP Fri Oct 9 13:04:14 AEST 2015 x86_64 x86_64 x86_64 GNU/Linux

**Comment by narqo** [comment]
Hey @Firstyear, I currently have an issue with similar signs. It looks like docker had run the command you passed within the `Dockerfile`, before IPv6 network was established. 

Could you check this two scenarios based on your description from the issue:

1\. Run `nmap` right after container was started:
~~~sh
› docker run -i --rm fedora:latest nmap -6 fe80::1%eth0 -p 3128
~~~ 

2\. Wait a little before running `nmap`
~~~sh
› docker run -i --rm fedora:latest sh -c 'sleep3; nmap -6 fe80::1%eth0 -p 3128'
~~~

My guess here is that everything will work as expected in the second case.

/cc @thaJeztah could you suggest anyone who could look at the issue?

**Comment by thaJeztah** [comment]
probably;

ping @aboch @mavenugo or @mrjana should be able to have a look :)

**Comment by narqo** [comment]
Some additional random thoughts I currently have: my host's `/etc/resolv.conf` has only one DNS record (which comes from my cloud provider). 

If I run:

```
› docker run ... bash -c 'strace -e poll,select,connect,recvfrom,sendto nc mirror.yandex.ru 80'
```

strace's log shows that there is a timeout in DNS resolve:

~~~
···
connect(3, {sa_family=AF_INET6, sin6_port=htons(53), inet_pton(AF_INET6, "2a02:6b8:0:3400::1023", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, 28) = 0
poll([{fd=3, events=POLLOUT}], 1, 0)    = 1 ([{fd=3, revents=POLLOUT}])
poll([{fd=3, events=POLLIN}], 1, 5000)  = 0 (Timeout)
poll([{fd=3, events=POLLOUT}], 1, 0)    = 1 ([{fd=3, revents=POLLOUT}])
poll([{fd=3, events=POLLIN}], 1, 5000)  = 0 (Timeout)
~~~

But if I add a small `sleep` as the first command, DNS will be resolved as expected:


~~~
› sudo docker run --rm -i test1 bash -c 'sleep 2; strace -e poll,select,connect,recvfrom,sendto nc mirror.yandex.ru 80'

···
connect(3, {sa_family=AF_INET6, sin6_port=htons(53), inet_pton(AF_INET6, "2a02:6b8:0:3400::1023", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, 28) = 0
poll([{fd=3, events=POLLOUT}], 1, 0)    = 1 ([{fd=3, revents=POLLOUT}])
poll([{fd=3, events=POLLIN}], 1, 5000)  = 1 ([{fd=3, revents=POLLIN}])
recvfrom(3, "\336\\\201\200\0\1\0\0\0\1\0\0\6mirror\6yandex\2ru\0\0\1"..., 2048, 0, {sa_family=AF_INET6, sin6_port=htons(53), inet_pton(AF_INET6, "2a02:6b8:0:3400::1023", &sin6_addr), sin6_flowinfo=0, sin6_scope_id=0}, [28]) = 108
poll([{fd=3, events=POLLIN}], 1, 4996)  = 1 ([{fd=3, revents=POLLIN}])
~~~

I'm pretty sure that everything is file with this named server as when I add the second host to the `resolv.conf`  (no matter where: at the top or at the bottom of the list) the same `nc` call works _only_ with second DNS host from the list of nameservers.

E.g. with resolv.conf

~~~
nameserver A
nameserver B
~~~

or 

~~~
nameserver B
nameserver A
~~~

the resolve call to the first nameserver will alway be timedout.

**Comment by thaJeztah** [comment]
Thanks for reporting @Firstyear  this *could* be the same as this issue; https://github.com/docker/docker/issues/18871, which should be resolved by https://github.com/docker/docker/pull/19198,and is part of the docker 1.10.0 release.

If you have a system available for testing (not on production yet), would you be able to give the current release candidate a spin and see if this is resolved for you? https://github.com/docker/docker/releases/tag/v1.10.0-rc1 please note that the 1.10 release uses a content-addressable storage, which requires a migration step the first time docker is run. Depending on the number of images and containers on your host, this migration may take a long time to complete. Read more on this in the wiki https://github.com/docker/docker/wiki/Engine-v1.10.0-content-addressability-migration

**Comment by thaJeztah** [comment]
ping @aboch can you have a look, do you think this is resolved by https://github.com/docker/docker/pull/19198?

**Comment by thaJeztah** [comment]
ping @Firstyear @narqo are you still having this issue on docker 1.10 or 1.11? Can we close this?

**Comment by Firstyear** [comment]
Seems to be fixed on 1.10 from what I can see.
